scikeras/wrappers.py

```diff
@@ -279,6 +345,12 @@ def _build_keras_model(self):
         # make serializable
         make_model_picklable(model)
 
+        # compile model if user gave us an un-compiled model
+        if not (hasattr(model, "loss") and hasattr(model, "optimizer")):
+            if compile_kwargs is None:
+                compile_kwargs = self._get_compile_kwargs()
+            model.compile(**compile_kwargs)
```

Comment:
What do you mean by not the correct value? I feel that we should let Keras/TF decide what is a valid value or not.

Reply:
> the correct value?

I mean "correct" as in "the user-passed `loss` to `BaseWrapper.__init__` matches the loss of the user-compiled model." i.e, if the user passes `loss="mse"` but `est.model_.loss == None`.

Reply:
Hmm I see. Is this a simple check? It would have to cover all cases within ~5 LOC. I don't think we want complex runtime checking, I'd rather have that live in the tests and then we assume that at runtime we compiled correctly (until someone submits a bug report showing otherwise)

Reply:
By "simple" I mean how would we account for the user passing "SGD" but we end up with a `tf.keras.optimizers.SGD` instance. That's not incorrect, but we also can't compare those values without just duplicating logic. Unless I am missing something?

Reply:
I think https://github.com/adriangb/scikeras/pull/66#discussion_r492863842 is relevant. Maybe it's not future work to implement `shorthands` if this use case is showing up.

I think checking names would suffice for now (i.e, only that `shorthands[self.model_.optimizer] == shorthands[self.optimizer]`, not learning rate/momentum/etc).

Reply:
Wouldn't we have to recursively check the names, ex to support `loss=["binary_crossentropy", ...]`?

When would this check fail (i.e. `shorthands[self.model_.optimizer] != shorthands[self.optimizer]`)? The only way I could see it failing is if we had a bug in our compilation code, but I may be missing something.

And what if Keras does some post-processing with the paramters? For example, they say that they map `metrics="accuracy"` to `BinaryAccuracy`, `CategoricalAccuracy` or `SparseCategoricalAccuracy` depending on the loss function ([link](https://www.tensorflow.org/api_docs/python/tf/keras/Model?version=nightly#compile)). I also recall times where Keras wraps the function you give it in strange ways (although I have to double check, I don't remember the specifics). I'm concerned that we wouldn't be able to reliably check`self.model_.{optimizer, loss, metrics}`. Which would leave us checking `shorthands[compile_kwargs["optimizer"] == shorthands[self.optimizer]`.

Reply:
> When would this check fail 

When the user doesn't compile their model correctly. Maybe they're doing some custom processing and are accidentally popping some key from `compile_kwargs` and using `model.compile(**compile_kwargs)`.

All I'm saying is do basic checks to make sure the user didn't make a boneheaded mistake like above. This code might  suffice:

``` python
class BaseWrapper:

    def _build_model():
        ...
        model = build_fn(self.model)
        if compiled(model)
            if shorthands[self.loss] != shorthands[self.model_.loss]:
                raise ValueError(...)
            if shorthands[self.optimizer] != shorthands[self.model_.optimizer]:
                raise ValueError(...)
```

Reply:
I see. That could indeed be valuable. So this is mainly to be nice to users and give them a descriptive SciKeras error instead of the alternative, which I think would be one of:
(1) confusing results after fitting/predicting
(2) a TensorFlow error ala `no gradients for output` that is confusing and not user friendly.

Am I understanding correctly?

What would we do in the following scenario:

```python3
def get_model():
    model = Model(...)
    model.compile(optimizer="sgd")  # or any other hardcoded optimizer

est = KerasClassifier(get_model)  # results in self.optimizer == None because that is the default
```
I feel that we should allow users to do this (in fact, if we do not, we would be breaking the old API carried over from the TF repo). I also think we should keep the default for `optimizer` as `None` because (1) there is no universal logical default and (2) that's the default for `Model.compile`.

Reply:
Yup, (1) is the case I'm worried about: that's when the user passed a value (and expects a behavior change) but that change is not reflected in the execution. Case (2) happen when `loss=None` and `fit` is called (which I think should also raise an error).

I think SciKeras should raise errors if users pass values to `BaseWrapper` but they're ignored. I don't think SciKeras should raise errors for hardcoded parameters.

Separately, I also think the documentation should return un-compiled models for the vast majority of cases (and definitely in the "getting started" guide).

Reply:
> I think SciKeras should raise errors if users pass values to `BaseWrapper` but they're ignored. I don't think SciKeras should raise errors for hardcoded parameters.

User-defined parameters (i.e. `KerasClassifier(.., myparam=1)`) already raise a TypeError if `get_model` does not accept a parameter by that name. Or are you referring to another case?

> Separately, I also think the documentation should return un-compiled models for the vast majority of cases (and definitely in the "getting started" guide).

Big +1 on this. I feel like this would really mitigate the issue in the first place. I'd assume that if users are trying to customize things to the point where they've read a good chunk of the docs and realized that SciKeras' compiling framework cannot accommodate their needs, they know that they are doing advanced stuff and that mistakes they make could lead to confusing results.

Reply:
> Or are you referring to another case?

Yes. I'm talking about the case when `est = KerasRegressor(..., optimizer="sgd")` but `est.model_.optimizer == "rmsprop"`. That's possible with this:

``` python
def model(compile_kwargs=None, meta=None):
    opt = compile_kwargs.pop("optimizer", "sgd")
    # custom processing w/ opt
    model.compile(**compile_kwargs)

est = KerasRegressor(model=model, optimizer="sgd")
print(est.model_.optimizer)  # rmsprop
```

As a user, I would call this unexpected behavior: I meant to pass `optimizer=opt` to `model.compile` but forgot. That error shouldn't pass silently.

Reply:
How would we test against that while still allowing the following, which IIRC is valid in the TF repo implementation:

```python3
def model(optimizer):
    model.compile(optimizer="hardcoded_optimizer")
    return model

est = KerasRegressor(model=model, optimizer="sgd")
```

Or would be decide to break that API?

Reply:
In any case, it seems to me that this could and should be handled in a separate PR. If you're okay with that let's move forward with this one and continue the discussion in #79

Reply:
Yeah, let's delegate to #79. I'll resolve this issue.

Reply:
Great! It looks like there are no more unresolved threads. Let me know when/if you think we are ready to merge.